packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -95,110 +89,131 @@ function patch(n1: VNode, n2: VNode) {
 
     switch (n2.type) {
         case VNodeType.Text:
-            patchText(n1 as VText, n2);
+            // VText has no special capability, fallback to the owner's renderer
+            patchText(n1 as VText, n2, renderer);
             break;
 
         case VNodeType.Comment:
-            patchComment(n1 as VComment, n2);
+            // VComment has no special capability, fallback to the owner's renderer
+            patchComment(n1 as VComment, n2, renderer);
             break;
 
         case VNodeType.Element:
-            patchElement(n1 as VElement, n2);
+            patchElement(n1 as VElement, n2, n2.data.renderer ?? renderer);
             break;
 
         case VNodeType.CustomElement:
-            patchCustomElement(n1 as VCustomElement, n2);
+            patchCustomElement(n1 as VCustomElement, n2, n2.data.renderer ?? renderer);
             break;
     }
 }
 
-export function mount(node: VNode, parent: ParentNode, anchor: Node | null) {
+export function mount(node: VNode, parent: ParentNode, renderer: RendererAPI, anchor: Node | null) {
     switch (node.type) {
         case VNodeType.Text:
-            mountText(node, parent, anchor);
+            // VText has no special capability, fallback to the owner's renderer
+            mountText(node, parent, anchor, renderer);
             break;
 
         case VNodeType.Comment:
-            mountComment(node, parent, anchor);
+            // VComment has no special capability, fallback to the owner's renderer
+            mountComment(node, parent, anchor, renderer);
             break;
 
         case VNodeType.Element:
-            mountElement(node, parent, anchor);
+            mountElement(node, parent, anchor, node.data.renderer ?? renderer);
```

Comment:
Although this new style kinda hides the key piece of this PR, I guess we can improve it with a comment like this.
```suggestion
            // If the vnode data has a renderer override use it, else fallback to owner's renderer
            mountElement(node, parent, anchor, node.data.renderer ?? renderer);
```